core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/etcs/ETCSBrakingCurves.kt

```diff
@@ -40,220 +45,213 @@ enum class BrakingType {
-    val indCurveEoa = Envelope.make(*indCurveEoaParts.toList().toTypedArray())
-    val indCurveSvl = Envelope.make(*indCurveSvlParts.toList().toTypedArray())
-
-    // The indication curve is the minimum of the SvL curve maintaining release speed and the EoA
-    // curve, between [minBeginPos, targetPosition]. We need the 2 envelopes to have the exact same
-    // position-range.
-    return minEnvelopes(indCurveEoa, indCurveSvl)
+    return res
 }
 
-/** Compute braking curves at every limit of authority. */
-fun addBrakingCurvesAtLOAs(
+/**
+ * Compute braking curves at every end of authority until their speed intersects with inputted
+ * envelope.
+ */
+fun computeBrakingCurvesAtEOAs(
     envelope: Envelope,
     context: EnvelopeSimContext,
-    limitsOfAuthority: Collection<LimitOfAuthority>
-): Envelope {
-    val sortedLimitsOfAuthority = limitsOfAuthority.sortedBy { it.offset }
-    val beginPos = envelope.beginPos
-    val maxSpeedEnvelope = envelope.maxSpeed
-    var envelopeWithLoaBrakingCurves = envelope
-    var builder = OverlayEnvelopeBuilder.forward(envelopeWithLoaBrakingCurves)
+    endsOfAuthority: Collection<EndOfAuthority>
+): List<Map<BrakingCurveType, ETCSBrakingCurve?>> {
+    val sortedEndsOfAuthority = endsOfAuthority.sortedBy { it.offsetEOA }
+    val res = mutableListOf<Map<BrakingCurveType, ETCSBrakingCurve?>>()
+    for (endOfAuthority in sortedEndsOfAuthority) {
+        res.add(computeEOABrakingCurves(endOfAuthority, context, envelope, 0.0))
+    }
+    return res
+}
 
-    for (limitOfAuthority in sortedLimitsOfAuthority) {
-        val targetPosition = limitOfAuthority.offset.distance.meters
-        assert(targetPosition > 0.0)
-        val targetSpeed = limitOfAuthority.speed
-        assert(targetSpeed > 0.0)
-        val fullIndicationCurve =
-            computeEbdFullIndicationCurve(context, targetPosition, targetSpeed, maxSpeedEnvelope)
-        val endOfIndicationCurve = fullIndicationCurve.endPos
-        assert(endOfIndicationCurve <= targetPosition)
-
-        val fullIndCurveWithMaintain: Envelope
-        if (endOfIndicationCurve < targetPosition) {
-            // Maintain target speed until target position, i.e. LOA.
-            val maintainTargetSpeedCurve =
-                EnvelopePart.generateTimes(
-                    listOf(EnvelopeProfile.CONSTANT_SPEED),
-                    doubleArrayOf(endOfIndicationCurve, targetPosition),
-                    doubleArrayOf(targetSpeed, targetSpeed)
-                )
-            fullIndCurveWithMaintain = Envelope.make(fullIndicationCurve, maintainTargetSpeedCurve)
-        } else {
-            fullIndCurveWithMaintain = Envelope.make(fullIndicationCurve)
-        }
+/** Compute LoA braking curves: compute EBD-based curves for LoA. */
+private fun computeLOABrakingCurves(
+    limitOfAuthority: LimitOfAuthority,
+    context: EnvelopeSimContext,
+    maxSpeedEnvelope: Envelope,
+    beginPos: Double
+): Map<BrakingCurveType, ETCSBrakingCurve?> {
```

Comment:
This is redundant on `BrakingCurveType`:
```suggestion
): Map<BrakingCurveType, Envelope?> {
```
Ditto for other uses of this Map.

Reply:
I know. It was a choice to make sure we weren't using the wrong type of curve. This'll disappear with Eckter's comment on using an EnumMap anyhow, so it'll be managed :)

Reply:
Yeah I'm gonna keep the redundancy, as I prefer having handling ETCSBrakingCurve objects than non-typed envelopes.

You suggested making `ETCSBrakingCurve` a `Pair<BrakingCurveType, Envelope>` to use it directly in the map, but that would make it so we get the type and envelope of the curve by using `.first()` and `.second()`. I don't quite like that ngl, I'd rather have what I did atm.